Feature/1174/introduce notification functionality

ryanwelcher opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses. -->
Addresses issue #1174 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

tofumatt commented on assets/js/googlesitekit/datastore/site/index.test.js:
```diff
@@ -58,4 +59,17 @@ describe( 'core/site store', () => {
+				'receiveNotifications',
+				'receiveNotificationsFailed',
+			];
+			expect( Object.keys( actions ) ).toEqual( expect.arrayContaining( actionsToExpect ) );
```
I know I'm guilty of suggesting this as a possible test when we spoke, but seeing it here I'm not sure it adds any value: this is essentially testing that the notifications API exports _at least_ these actions, even if our code never uses them. The notifications store is a public API for us and we should be sure to maintain it, but that's not what we should be testing here.

If the notifications API were to change we'd need to change this test… I'd rather a more "functional"/integration test that _used_ the notifications actions be included here so we'd be alerted to breaking changes in the notifications API.

-----

I'm sort of having a back-and-forth here and posting my thoughts, because ultimately I guess it's good to test this and make sure the module is behaving as expected. So we can leave this here, but I'm leaving the above waffling in case it's useful to have on record 😆 

Since there's a `notifications.js` file, though, these tests should go into a `notifications.test.js` file. I know it's a pain/fair bit of extra boilerplate, but we should keep tests as close to what they're testing as possible 😄

ryanwelcher commented:
@tofumatt thanks for your input on this! I've moved the tests into their own file and added some more tests to ensure that the action being dispatched matches the API. I know that `createNotificationsStore` is heavily tested but this should warn us if there are changes to the object being sent with the action.

tofumatt commented on assets/js/googlesitekit/datastore/site/notifications.test.js:
```diff
@@ -0,0 +1,85 @@
+			payload: { notification },
+			type: ADD_NOTIFICATION,
+		};
+		expect( actions.addNotification( notification ) ).toEqual( expectedAction );
```
So, I get the intention of checking the actions here but again: if the internal implementation of how notifications dispatch actions changes we'll have to update the tests here even though they aren't related. This is testing how notifications dispatch their actions instead of _that_ our store has notifications. If we change the `addNotification` action creator by adding a new attribute to the payload, it shouldn't affect this datastore as the notification part of the store will handle things.

But we'd have to update this test even though nothing about this store or the functions it used would change externally. So I think it's best to just have the `has appropriate notification methods` test for this file; testing the individual actions and especially their payloads (which are not consumed by reducers outside of the notifications store—they're essentially "private"/internal) is fragile and doesn't provide much extra benefit to this store.

I think it's best to remove these and just leave the "has appropriate notification methods" one. Sorry if my previous suggestion was confusing!

ryanwelcher replied:
@tofumatt I've removed the tests. Thanks for your feedback!
